### ISSUE TITLE
Finetuning the moe model Qwen3vl-30b-a3b encountered CUDA OOM.

### ISSUE THREAD

**Issue by yzxyzh** [QwenLM/Qwen3-VL#1565]
Hello Qwen teams, thanks for your wonderful works!

Currently I used your scripts for finetuning Qwen3vl-30b-a3b model with my custom video data. the arguments i adopt will be pasted at the end of this post. And unfortunately I had encountered cuda oom problem during finetuning.

My machine has 8 gpus, and the gpu is H20-141GB, and I received cuda oom error when i set the --model_max_length parameter to 4096 and 2048. And the code I was using was exactly the latest repo you provided. The OOM happens at gradient backprop stage according to the logs. when the finetuning started, all cards had around 100GB memories occupied, and the program crashed then reported cuda OOM when the backprop started.

Could you please provide a "default" setting of training devices (GPU numbers and their memories) that in your case would run the finetuning process for Qwen3vl-30b-a3b successfully, or could the OOM be a bug?

BTW: there might be several bugs in your repo, below are details:

issue 1:

the following model selection code snippet in train_qwen.py : 

    if "qwen3" in model_args.model_name_or_path.lower() and "moe" in model_args.model_name_or_path.lower():
        model = Qwen3VLMoeForConditionalGeneration.from_pretrained(
            model_args.model_name_or_path,
            cache_dir=training_args.cache_dir,
            attn_implementation=attn_implementation,
            dtype=(torch.bfloat16 if training_args.bf16 else None),
        )
        data_args.model_type = "qwen3vl"
    elif "qwen3" in model_args.model_name_or_path.lower():
        model = Qwen3VLForConditionalGeneration.from_pretrained(
            model_args.model_name_or_path,
            cache_dir=training_args.cache_dir,
            attn_implementation=attn_implementation,
            dtype=(torch.bfloat16 if training_args.bf16 else None),
        )

for qwenvl-30b-a3b, since the name does not contain "moe", the default selection would be Qwen3VLForConditionalGeneration in this case, which i guess would be wrong? so i modified the 'moe' to 'a3b' and the code then selects Qwen3VLMoeForConditionalGeneration

issue 2:

the following code snippet in data_processor.py:

 if "video_grid_thw" in data_dict:
            video_grid_thw = data_dict.get("video_grid_thw")
            if not isinstance(video_grid_thw, Sequence):
                video_grid_thw = [video_grid_thw]
            second_per_grid_ts = [
                self.data_args.processor.video_processor.temporal_patch_size
                / self.data_args.processor.video_processor.fps
            ] * len(video_grid_thw)
        else:

there were no data_args.processor so i guess it would be " self.processor...." instead?

And finally my sft settings: (modified from your sft_30a3b.sh)

# Training arguments
args="
    --deepspeed ${deepspeed} \
    --model_name_or_path "${llm}" \
    --dataset_use ${datasets} \
    --data_flatten True \
    --tune_mm_vision False \
    --tune_mm_mlp True \
    --tune_mm_llm True \
    --bf16 \
    --output_dir ${output_dir} \
    --num_train_epochs 3 \
    --per_device_train_batch_size ${batch_size} \
    --per_device_eval_batch_size $((batch_size*2)) \
    --gradient_accumulation_steps ${grad_accum_steps} \
    --max_pixels 50176 \
    --min_pixels 784 \
    --video_fps 1 \
    --video_max_frames 10 \
    --video_min_frames 2 \
    --video_max_pixels 1164240 \
    --video_min_pixels 200704 \
    --eval_strategy "no" \
    --save_strategy "steps" \
    --save_steps 1000 \
    --save_total_limit 1 \
    --learning_rate ${lr} \
    --weight_decay 0 \
    --warmup_ratio 0.03 \
    --max_grad_norm 1 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --run_name ${run_name} \
    --report_to none"

Looking forward for your replies!

**Comment by John-Ge** [comment]
Thank you! We have fixed the problem in [PR](https://github.com/QwenLM/Qwen3-VL/pull/1570).

**Comment by yzxyzh** [comment]
> Thank you! We have fixed the problem in [PR](https://github.com/QwenLM/Qwen3-VL/pull/1570).

with pleasure! And how about the OOM problem I've mentioned? would you please provide a default hardware setting that can run the training scripts for the qwen3-vl moe model without encountering any OOM issue? 

For a quick fix for myself, I tried to add this line in your zero_2.json according to the suggestions given by gpt5:

"offload_optimizer": { "device": "cpu", "pin_memory": true },

and the script now trains ok on my machine, but the speed is rather slow. Still i wound like to know the minimum hardware that could run the training using the default zero2 settings provided by your team so that i could have an idea of how many additional machines would i require at minimum to run the training, or could there be a setting error of me.

**Comment by John-Ge** [comment]
I use zero2 on 32 GPUs. You could set the total batch size to 128 or larger if you have enough memory.

**Comment by yzxyzh** [comment]
> I use zero2 on 32 GPUs. You could set the total batch size to 128 or larger if you have enough memory.

Thanks for your clarification! I'll try myself.

**Comment by luckyrz0** [comment]
> > I use zero2 on 32 GPUs. You could set the total batch size to 128 or larger if you have enough memory.我在 32 个 GPU 上使用 zero2。如果您有足够的内存，您可以将总批次大小设置为 128 或更大。
> 
> Thanks for your clarification! I'll try myself.感谢您的澄清！我会自己试试。

Hello, have you solved this problem yet
